lightning/src/ln/channelmanager.rs

```diff
@@ -109,71 +109,119 @@ use crate::ln::script::ShutdownScript;
-		/// payment was sent to a phantom node (one hop beyond the current node), but can be
-		/// settled by this node.
+		/// If the onion had forwarding instructions to one of our phantom node SCIDs, this will
+		/// provide the onion shared secret used to decrypt the second level of forwarding
+		/// instructions.
 		phantom_shared_secret: Option<[u8; 32]>,
-		/// See [`RecipientOnionFields::custom_tlvs`] for more info.
+		/// Custom TLVs which were set by the sender.
+		///
+		/// For HTLCs received by LDK, these will ultimately bubble back up as
+		/// [`RecipientOnionFields::custom_tlvs`].
 		custom_tlvs: Vec<(u64, Vec<u8>)>,
 	},
-	/// Incoming keysend (sender provided the preimage in a TLV).
+	/// The onion indicates that this is for payment to us but which contains the preimage for
+	/// claiming included, and is unrelated to any invoice we'd previously generated (aka a
+	/// "keysend" or "spontaneous" payment).
 	ReceiveKeysend {
-		/// This was added in 0.0.116 and will break deserialization on downgrades.
+		/// Information about the amount the sender intended to pay and possibly a token to
+		/// associate MPP parts of a larger payment.
+		///
+		/// This will only be filled in if receiving MPP keysend payments is enabled, and it being
+		/// present may break downgrades to versions of LDK prior to 0.0.116.
 		payment_data: Option<msgs::FinalOnionHopData>,
 		/// Preimage for this onion payment. This preimage is provided by the sender and will be
 		/// used to settle the spontaneous payment.
 		payment_preimage: PaymentPreimage,
-		/// See [`RecipientOnionFields::payment_metadata`] for more info.
+		/// Additional data which we (allegedly) instructed the sender to include in the onion.
+		///
+		/// For HTLCs received by LDK, this will ultimately bubble back up as
+		/// [`RecipientOnionFields::payment_metadata`].
 		payment_metadata: Option<Vec<u8>>,
 		/// CLTV expiry of the received HTLC.
 		/// Used to track when we should expire pending HTLCs that go unclaimed.
 		incoming_cltv_expiry: u32,
-		/// See [`RecipientOnionFields::custom_tlvs`] for more info.
+		/// Custom TLVs which were set by the sender.
+		///
+		/// For HTLCs received by LDK, these will ultimately bubble back up as
+		/// [`RecipientOnionFields::custom_tlvs`].
 		custom_tlvs: Vec<(u64, Vec<u8>)>,
 	},
 }
 
-/// Full details of an incoming HTLC, including routing info.
+/// Information about an incoming HTLC, including the [`PendingHTLCRouting`] describing where it
+/// should go next.
 #[derive(Clone)] // See Channel::revoke_and_ack for why, tl;dr: Rust bug
 pub struct PendingHTLCInfo {
 	/// Further routing details based on whether the HTLC is being forwarded or received.
 	pub routing: PendingHTLCRouting,
-	/// Shared secret from the previous hop.
-	/// Used encrypt failure packets in the event that the HTLC needs to be failed backwards.
+	/// The onion shared secret we build with the sender used to decrypt the onion.
```

Comment:
Though “sender” makes perfect sense in the context, we use “sender” usually for the original creator of the payment. Would “previous hop” be a better phrasing here?

```suggestion
	/// The onion shared secret we build with the previous hop used to decrypt the onion.
```

Reply:
I've been trying to get my head around this concept, and after a bit of digging, I found some relevant info in [Bolt #4](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#returning-errors).

It reads:
> Intermediate hops store the shared secret from the forward path and reuse it to obfuscate any corresponding return packet during each hop. In addition, each node locally stores data regarding its own sending peer in the route, so it knows where to return forward any eventual return packets.

AFAIU, this talks about creating a secret with the own sending peer (adjacent peer) and not with **The Sender**.

So, can you help me point to a resource I can read up to understand why and how we create keys with **The Sender** and do so in a way that doesn't compromise privacy?

Thanks a lot!